core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1387,10 +1388,10 @@ public void startRecovery(DiscoveryNode localNode, DiscoveryNode sourceNode, Rec
             IndexMetaData indexMetaData = indexSettings().getIndexMetaData();
             Index mergeSourceIndex = indexMetaData.getMergeSourceIndex();
             if (mergeSourceIndex != null && shardRouting.allocatedPostIndexCreate(indexMetaData) == false && shardRouting.primary()) {
-                List<IndexShard> startedShards = new ArrayList<>();
-                IndexService sourceIndexService = indicesService.indexService(mergeSourceIndex);
-                IndexMetaData metaData = sourceIndexService.getMetaData();
-                int numShards = metaData != null ? metaData.getNumberOfShards() : -1;
+                final List<IndexShard> startedShards = new ArrayList<>();
+                IndexMetaData sourceIndexMetaData = indicesService.getIndexMetaData(mergeSourceIndex);
```

Comment:
can we get the metaData from IndexService.getIndexSettings? This will mean IndicesService doesn't need to proxy clusterService.state()